include/ftl/fibtex.h

```diff
@@ -0,0 +1,120 @@
+/**
+ * FiberTaskingLib - A tasking library that uses fibers for efficient task switching
+ *
+ * This library was created as a proof of concept of the ideas presented by
+ * Christian Gyrling in his 2015 GDC Talk 'Parallelizing the Naughty Dog Engine Using Fibers'
+ *
+ * http://gdcvault.com/play/1022186/Parallelizing-the-Naughty-Dog-Engine
+ *
+ * FiberTaskingLib is the legal property of Adrian Astley
+ * Copyright Adrian Astley 2015 - 2018
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "task_scheduler.h"
+#include "atomic_counter.h"
+
+
+namespace ftl {
+/**
+ * A fiber aware mutex. Does not block in the traditional way.
+ */
+class Fibtex {
+public:
+	/**
+	 * All Fibtex's have to be aware of the task scheduler in order to yield.
+	 *
+	 * @param taskScheduler ftl::TaskScheduler that will be using this mutex.
+	 */
+	explicit Fibtex(ftl::TaskScheduler *taskScheduler)
+		: m_taskScheduler(taskScheduler),
+		  m_atomicCounter(taskScheduler, 0),
+		  m_ableToSpin(taskScheduler->GetThreadCount() > 1) {}
```

Comment:
Can you explain the purpose of `m_ableToSpin` ?

In all the usages, it looks like just an extra operation. The `else` code does the exact same thing as the `if not` code

Reply:
Basically, with a spin lock and under a single backing thread you will just spin forever as there is no other fiber that will ever free the lock. This is eating the cost of an easily predicted branch in exchange for dealing with that case.

Reply:
But if there is only one thread, how could it ever spin in the first place? There is no contention, so the lock should always succeed before hitting the Wait().

Reply:
There is a theoretically possible situation where this mutex is locked by fiber 0, then queues a job and waits on a counter. While this mutex is still locked, some other thread comes along and tries to lock it, causing a deadlock.

Reply:
Ahhh. I see now